lib/drum.hoon

```diff
@@ -698,10 +698,115 @@
+    |=  a/@  ^-  ?
+    ?|  &((gte a '0') (lte a '9'))
+        &((gte a 'A') (lte a 'Z'))
+        &((gte a 'a') (lte a 'z'))
+    ==
+  ::
+  ++  next-edge
+    |=  buf/(list @c)
+    =|  i/@ud
+    =+  m=|
+    |-  ^+  i
+    ?~  buf  i
+    =+  w=(alnum i.buf)
+    =.  m  |(m w)
+    ?:  &(m !|(=(0 i) w))
+      i
+    $(i +(i), buf t.buf)
+  ::
+  ++  next-word
+    |=  buf/(list @c)
+    =|  i/@ud
+    |-  ^+  i
+    ?:  |(?=($~ buf) (alnum i.buf))
+      i
+    $(i +(i), buf t.buf)
+  ::
+  ++  jump-bwrd
+    |=  {pos/@ud buf/(list @c)}
+    ^-  @ud
+    (sub pos (next-edge (flop (scag pos buf))))
+  ::
+  ++  jump-fwrd
+    |=  {pos/@ud buf/(list @c)}
+    ^-  @ud
+    (add pos (next-edge (slag pos buf)))
+  ::
   ++  ta-met                                          ::  meta key
     |=  key/@ud
-    ~&  [%ta-met key]
-    +>
+    ?+    key    ta-bel
+      $dot  ?.  &(?=(^ old.hit) ?=(^ -.old.hit))
+              ta-bel
+            =+  old=`(list @c)`-.old.hit
+            =+  b=(jump-bwrd (dec (lent old)) old)
+            %-  ta-hom(ris ~)
+            (ta-cat pos.inp (slag b old))
+            ::
+      $bac  ?:  =(0 pos.inp)
+              ta-bel
+            =+  b=(jump-bwrd pos.inp buf.say.inp)
+            %-  ta-hom(kil `(slag b (scag pos.inp buf.say.inp)), ris ~)
+            (ta-cut b (sub pos.inp b))
+            ::
+      $b    ?:  =(0 pos.inp)
+              ta-bel
+            +>(pos.inp (jump-bwrd pos.inp buf.say.inp))
+            ::
+      $c    ?:  =(pos.inp (lent buf.say.inp))
+              ta-bel
+            =+  sop=(add pos.inp (next-word (slag pos.inp buf.say.inp)))
+            =+  ^=  prod
+            %-  ta-hom
+            :~  %mor
+              [%del sop]
+              :+  %ins  sop
+              (turf (cuss [(tuft (snag sop buf.say.inp))]~))
+            ==
+            prod(pos.inp (jump-fwrd sop buf.say.inp))
+            ::
+      $d    ?:  =(pos.inp (lent buf.say.inp))
+              ta-bel
+            =+  f=(jump-fwrd pos.inp buf.say.inp)
+            %-  ta-hom(kil `(slag pos.inp (scag f buf.say.inp)), ris ~)
+            (ta-cut pos.inp (sub f pos.inp))
+            ::
+      $f    ?:  =(pos.inp (lent buf.say.inp))
+              ta-bel
+            +>(pos.inp (jump-fwrd pos.inp buf.say.inp))
+            ::
+      $r    ta-bel
+      $t    =+  a=(jump-fwrd pos.inp buf.say.inp)
+            =+  b=(jump-bwrd a buf.say.inp)
+            =+  c=(jump-bwrd b buf.say.inp)
+            ?:  =(b c)
+              ta-bel
+            =+  prev=`(pair @ud @ud)`[c (jump-fwrd c buf.say.inp)]
+            =+  next=`(pair @ud @ud)`[b a]
+            =+  ^=  prod
+            %-  ta-hom
+            :~  %mor
+              (ta-cut p.next (sub q.next p.next))
+              (ta-cat p.next (slag p.prev (scag q.prev buf.say.inp)))
+              (ta-cut p.prev (sub q.prev p.prev))
+              (ta-cat p.prev (slag p.next (scag q.next buf.say.inp)))
+            ==
+            prod(pos.inp q.next)
+            ::
```

Comment:
This is awkward, but lets me order the edits in a straightforward manner. The alternative includes a lot more offset calculation...